File: agent/tools/MCP/npm/mcp-1panel/src/tools/runtimes.js
/**
 * mcp-1panel — Runtimes 模块工具（12 个）
 */
import { tool, toolWithParams, fmtObj, fmtList } from "./helpers.js";
import { api } from "../api-proxy.js";

export const RUNTIMES_TOOLS = {

  panel_runtimes: tool("运行环境列表 — PHP/Node 等", async () => {
    const d = await api.post("/runtimes/search", { page: 1, pageSize: 50 });
    const items = d.items ?? [];
    const lines = [`= 运行环境 (共${d.total ?? items.length}个) =`];
    for (const r of items) lines.push(`  ${r.name ?? "?"}  ${r.type ?? "?"}  v${r.version ?? "?"}  [${r.status ?? "?"}]`);
    if (!items.length) lines.push("  (空)");
    return lines.join("\n");
  }),

  panel_runtime_detail: toolWithParams("运行环境详情", async (args) => {
    const d = await api.get(`/runtimes/${args.id}`);
    return fmtObj("运行环境详情", d);
  }, { id: { type: "number", default: 1 } }),

  panel_php_extensions: toolWithParams("PHP 扩展列表", async (args) => {
    const d = await api.get(`/runtimes/php/${args.id}/extensions`);
    return fmtList("PHP 扩展", Array.isArray(d) ? d : [], (e) => `${e.name ?? "?"}  ${e.version ?? ""}  ${e.enable ? "✅" : "⬜"}`);
  }, { id: { type: "number", default: 1 } }),

  panel_php_config: toolWithParams("PHP 运行配置", async (args) => {
    const d = await api.get(`/runtimes/php/config/${args.id}`);
    return fmtObj("PHP 配置", d);
  }, { id: { type: "number", default: 1 } }),

  panel_php_container: toolWithParams("PHP 容器配置", async (args) => {
    const d = await api.get(`/runtimes/php/container/${args.id}`);
    return fmtObj("PHP 容器配置", d);
  }, { id: { type: "number", default: 1 } }),

  panel_php_fpm_config: toolWithParams("PHP-FPM 配置", async (args) => {
    const d = await api.get(`/runtimes/php/fpm/config/${args.id}`);
    return fmtObj("PHP-FPM 配置", d);
  }, { id: { type: "number", default: 1 } }),

  panel_php_fpm_status: toolWithParams("PHP-FPM 运行状态", async (args) => {
    const d = await api.get(`/runtimes/php/fpm/status/${args.id}`);
    return fmtObj("PHP-FPM 状态", d);
  }, { id: { type: "number", default: 1 } }),

  panel_php_file: toolWithParams("PHP 配置文件内容", async (args) => {
    const d = await api.post("/runtimes/php/file", { id: args.id });
    return typeof d === "string" ? `= PHP 配置文件 =\n${d.slice(0, 2000)}` : fmtObj("PHP 配置文件", d, 200);
  }, { id: { type: "number", default: 1 } }),

  panel_node_modules: toolWithParams("Node.js 模块列表", async (args) => {
    const d = await api.post("/runtimes/node/modules", { id: args.id });
    return fmtList("Node 模块", Array.isArray(d) ? d.slice(0, 50) : [], (m) => `${m.name ?? "?"}  v${m.version ?? "?"}`);
  }, { id: { type: "number", default: 1 } }),

  panel_node_package: toolWithParams("Node.js 包脚本", async (args) => {
    const d = await api.post("/runtimes/node/package", { id: args.id });
    return fmtObj("Node 包脚本", d);
  }, { id: { type: "number", default: 1 } }),

  panel_supervisor_process_detail: toolWithParams("Supervisor 进程详情", async (args) => {
    const d = await api.get(`/runtimes/supervisor/process/${args.id}`);
    return fmtObj("Supervisor 进程", d);
  }, { id: { type: "number", default: 1 } }),

  panel_runtime_php_extensions_list: toolWithParams("PHP 可用扩展列表", async (args) => {
    const d = await api.post("/runtimes/php/extensions/search", { page: 1, pageSize: 50, runtimeId: args.id });
    const items = d.items ?? d ?? [];
    return fmtList("PHP 可用扩展", Array.isArray(items) ? items : [], (e) => `${e.name ?? "?"}  ${e.version ?? ""}  ${e.installed ? "✅" : "⬜"}`);
  }, { id: { type: "number", default: 1 } }),

};

File: agent/tools/MCP/mcp-1panel/src/mcp_1panel/tools/runtimes.py
"""MCP tools for runtimes (PHP/Node/Supervisor) via 1Panel API."""

from typing import Any, Callable, Dict, List
from mcp_1panel.tools.helpers import (
    icon_green,
    icon_red,
    icon_lock,
    icon_unlock,
    icon_status,
    header,
    fmt_val,
    fmt_obj,
    fmt_list,
    fmt_search,
    fmt_generic,
)


def register_tools(mcp, get_client, handlers=None):
    """Register all runtime-related MCP tools."""

    # ------------------------------------------------------------------ #
    # Runtime list (paginated search)
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_runtimes() -> str:
        """运行环境列表 — PHP/Node 等"""
        p = get_client()
        data = p.post("/runtimes/search", {"page": 1, "pageSize": 20})
        lines, items = fmt_search(data, "运行环境")
        for r in items:
            name = r.get("name", "?")
            rtype = r.get("type", "?")
            ver = r.get("version", "?")
            st_val = r.get("status", "")
            st = icon_status(st_val == "running" or st_val == "1") if st_val else ""
            lines.append(f"  {st} {name}  {rtype}  v{ver}  [{st_val}]")
        if not items:
            lines.append("  (空)")
        return "\n".join(lines)

    # ------------------------------------------------------------------ #
    # Runtime detail
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_runtime_detail(id: int) -> str:
        """运行环境详情"""
        p = get_client()
        data = p.get("/runtimes/detail", {"id": id})
        return fmt_generic(data, f"运行环境详情 (ID: {id})")

    # ------------------------------------------------------------------ #
    # PHP extensions (status)
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_php_extensions(id: int) -> str:
        """PHP 扩展列表"""
        p = get_client()
        data = p.get("/runtimes/php/extensions", {"id": id})
        lines = [header(f"PHP 扩展 (ID: {id})")]
        if isinstance(data, list):
            for ext in data:
                name = ext.get("name", ext.get("extension", str(ext)))
                loaded = ext.get("loaded", ext.get("status", ""))
                if loaded:
                    st = icon_green() if str(loaded).lower() in ("1", "true", "yes", "enabled") else icon_red()
                    lines.append(f"  {st} {name}")
                else:
                    lines.append(f"  {name}")
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    # ------------------------------------------------------------------ #
    # PHP config
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_php_config(id: int) -> str:
        """PHP 运行配置"""
        p = get_client()
        data = p.get("/runtimes/php/config", {"id": id})
        return fmt_generic(data, f"PHP 运行配置 (ID: {id})")

    # ------------------------------------------------------------------ #
    # PHP container
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_php_container(id: int) -> str:
        """PHP 容器配置"""
        p = get_client()
        data = p.get("/runtimes/php/container", {"id": id})
        return fmt_generic(data, f"PHP 容器配置 (ID: {id})")

    # ------------------------------------------------------------------ #
    # PHP-FPM config
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_php_fpm_config(id: int) -> str:
        """PHP-FPM 配置"""
        p = get_client()
        data = p.get("/runtimes/php/fpm/config", {"id": id})
        return fmt_generic(data, f"PHP-FPM 配置 (ID: {id})")

    # ------------------------------------------------------------------ #
    # PHP-FPM status
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_php_fpm_status(id: int) -> str:
        """PHP-FPM 运行状态"""
        p = get_client()
        data = p.get("/runtimes/php/fpm/status", {"id": id})
        return fmt_generic(data, f"PHP-FPM 状态 (ID: {id})")

    # ------------------------------------------------------------------ #
    # PHP file (config file content)
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_php_file(id: int) -> str:
        """PHP 配置文件内容"""
        p = get_client()
        data = p.get("/runtimes/php/file", {"id": id})
        return fmt_generic(data, f"PHP 配置文件 (ID: {id})")

    # ------------------------------------------------------------------ #
    # Node modules
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_node_modules(id: int) -> str:
        """Node.js 模块列表"""
        p = get_client()
        data = p.get("/runtimes/node/modules", {"id": id})
        lines = [header(f"Node.js 模块 (ID: {id})")]
        if isinstance(data, list):
            for mod in data:
                name = mod.get("name", mod.get("module", str(mod)))
                ver = mod.get("version", "")
                if ver:
                    lines.append(f"  {name}  v{ver}")
                else:
                    lines.append(f"  {name}")
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    # ------------------------------------------------------------------ #
    # Node package (package.json scripts)
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_node_package(id: int) -> str:
        """Node.js 包脚本 (package.json)"""
        p = get_client()
        data = p.get("/runtimes/node/package", {"id": id})
        return fmt_generic(data, f"Node.js package.json (ID: {id})")

    # ------------------------------------------------------------------ #
    # Supervisor process detail
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_supervisor_process_detail(id: int) -> str:
        """Supervisor 进程详情"""
        p = get_client()
        data = p.get("/runtimes/supervisor/process/detail", {"id": id})
        return fmt_generic(data, f"Supervisor 进程详情 (ID: {id})")

    # ------------------------------------------------------------------ #
    # PHP available extensions list
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_runtime_php_extensions_list(id: int) -> str:
        """PHP 可用扩展列表"""
        p = get_client()
        data = p.get("/runtimes/php/extensions/list", {"id": id})
        lines = [header(f"PHP 可用扩展 (ID: {id})")]
        if isinstance(data, list):
            for ext in data:
                name = ext.get("name", ext.get("extension", str(ext)))
                installed = ext.get("installed", ext.get("loaded", ""))
                if installed and str(installed).lower() in ("1", "true", "yes", "installed"):
                    st = icon_green()
                    lines.append(f"  {st} {name} (已安装)")
                else:
                    lines.append(f"  {name}")
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    # -- 收集 handler 供 resources 复用 --
    if handlers is not None:
        for _name, _fn in list(locals().items()):
            if _name.startswith("panel_") and callable(_fn):
                handlers[_name] = _fn
